src/ucp/core/ucp_am.c

```diff
@@ -629,26 +629,11 @@ static void ucp_am_zcopy_req_complete(ucp_request_t *req, ucs_status_t status)
 
 void ucp_am_zcopy_completion(uct_completion_t *self)
 {
-    ucp_request_t *req  = ucs_container_of(self, ucp_request_t,
-                                           send.state.uct_comp);
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t,
+                                          send.state.uct_comp);
 
-    if (req->send.state.dt.offset == req->send.length) {
+    if (ucp_request_send_state_check_comp(req)) {
         ucp_am_zcopy_req_complete(req, self->status);
-    } else if (self->status != UCS_OK) {
-        ucs_assert(self->status != UCS_INPROGRESS);
-
-        /* Avoid double release of resources */
-        req->send.state.uct_comp.func = NULL;
-
-        /* NOTE: the request is in pending queue if data was not completely sent,
-         *       just release resources by ucp_am_zcopy_req_complete() here and
-         *       complete request on purge pending later.
-         * TODO: - Make sure status contains the latest error happened on the
-         *         lane (to be supported in UCT)
-         *       - Revise multi-rail support
-         *       - Prevent other fragments to be sent
-         */
-        ucp_am_zcopy_complete_common(req);
```

Comment:
why removed?

Reply:
this is not needed anymore - we rely on a transport to complete all outstanding operations with error
there are two situations:
1. operation completed with error (it is checked in `ucp_request_send_state_check_comp()` now) and its Zcopy completion will release the UCP requests and invoke user's callback.
2. operation didn't complete yet, but UCP is doing pending_purge for UCT EP - here we check:
- if operation with Zcopy completion callback:
a) if the completion counter is 0, i.e. operation is not owned by UCT, complete it by invoking Zcopy completion callback manually;
b) if the completion counter is not 0, i.e. operation is owned by UCT, do nothing and wait when UCT returns this operation with error.
- if operation without Zcopy completion callback, just release UCP request.